src/main/java/org/openrewrite/java/migrate/trait/ScopedVariable.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2024 the original author or authors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.java.migrate.trait;
+
+import lombok.RequiredArgsConstructor;
+import lombok.Value;
+import org.jspecify.annotations.Nullable;
+import org.openrewrite.Cursor;
+import org.openrewrite.java.tree.J;
+import org.openrewrite.trait.SimpleTraitMatcher;
+import org.openrewrite.trait.Trait;
+
+@Value
+public class ScopedVariable implements Trait<J.VariableDeclarations.NamedVariable> {
+    Cursor cursor;
+    Cursor scope;
+    J.Identifier identifier;
+
+    @RequiredArgsConstructor
+    public static class Matcher extends SimpleTraitMatcher<ScopedVariable> {
+
+        @Override
+        protected @Nullable ScopedVariable test(Cursor cursor) {
+            if (cursor.getValue() instanceof J.VariableDeclarations.NamedVariable) {
+                J.VariableDeclarations.NamedVariable variable = cursor.getValue();
+                return new ScopedVariable(cursor, variable.getDeclaringScope(cursor), variable.getName());
```

Comment:
This `getDeclaringScope()` utility method is probably a good starting point. But to me it looks like it misses cases like try-with-resources and exception handlers, for example. Also note that the flow scoping of instanceof patterns is quite different and probably a bad fit. Maybe worth thinking about at least.